### PR TITLE
Supporting upgrade

### PR DESCRIPTION
**Pull request by bardielle**

### CODE REVIEW

**Comment by pkliczewski** [comment]
@bardielle please rebase

File: greenboot-health-check.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+if [ -x /usr/local/libexec/yggdrasil/device-worker ]; then
```
Comment (masayag) [inline greenboot-health-check.sh:3]: Can we check the status of the yggdrasild, nft and podman services which are required for the k4e-agent to operate ?

File: r4e-img-upgrade-iso.sh [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+# read input
+export IMAGE_NAME=$1
+export OSTREE_COMMIT=$2
+export IMAGE_VERSION=$3
+export NEW_IMAGE_NAME=${IMAGE_NAME}-${IMAGE_VERSION}
+
+# prepare local variables
+IMAGE_FOLDER=/var/www/html/${NEW_IMAGE_NAME}
+BLUEPRINT_FILE=blueprint.toml
+BLUPRINT_NAME=${IMAGE_NAME}
```
Comment (masayag) [inline r4e-img-upgrade-iso.sh:31]: s/BLUPRINT_NAME/BLUEPRINT_NAME

File: r4e-img-upgrade-iso.sh [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+# read input
+export IMAGE_NAME=$1
+export OSTREE_COMMIT=$2
+export IMAGE_VERSION=$3
+export NEW_IMAGE_NAME=${IMAGE_NAME}-${IMAGE_VERSION}
+
+# prepare local variables
+IMAGE_FOLDER=/var/www/html/${NEW_IMAGE_NAME}
+BLUEPRINT_FILE=blueprint.toml
```
Comment (masayag) [inline r4e-img-upgrade-iso.sh:30]: would we like to add an example of the upgrade blueprint ? 
perhaps to require it as a separate file, e.g. upgrade-blueprint.toml not to mix with the one which is expected to create the initial image ?

File: r4e-img-upgrade-iso.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+BLUPRINT_NAME=${IMAGE_NAME}
+
+
+# create a new blueprint
+composer-cli blueprints push $BLUEPRINT_FILE
+
+# validate blueprint
+echo "validating blueprint $BLUPRINT_NAME"
+composer-cli blueprints depsolve $BLUPRINT_NAME
```
Comment (masayag) [inline r4e-img-upgrade-iso.sh:39]: Does the script fail in case of a missing dependency ?

File: r4e-img-upgrade-iso.sh [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+# build image using the composer based on the parent commit create a new commit- first find the parent commit hash in http://<host-ip>/<image-name>/compose.json under "ostree-commit"
+echo "creating a new commit from parrent $OSTREE_COMMIT using blueprint $BLUPRINT_NAME"
+composer-cli compose start-ostree $BLUPRINT_NAME rhel-edge-commit --parent $OSTREE_COMMIT > temp_upgrade.out
```
Comment (masayag) [inline r4e-img-upgrade-iso.sh:43]: `rhel-edge-commit` is deprecated. we should switch to 8.5 and replace it with `edge-commit`.

File: r4e-img-upgrade-iso.sh [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+BUILD_ID=`cat temp_upgrade.out | awk '{print $2}'`
```
Comment (masayag) [inline r4e-img-upgrade-iso.sh:44]: there is an extra pipe here, could be simplified to: 
```
BUILD_ID=$(awk '{print $2}' temp_upgrade.out)
```

File: r4e-img-upgrade-iso.sh [hunk 7]
```diff
@@ -0,0 +1,59 @@
+# script for creating an update ISO
```
Comment (masayag) [inline r4e-img-upgrade-iso.sh:1]: the script is missing the ISO part ;-)

File: r4e-img-upgrade-iso.sh [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+while [ "`composer-cli compose status | grep $BUILD_ID | grep -c RUNNING`" != "0" ] ; do sleep 5 ; done
+if [ "`composer-cli compose status | grep $BUILD_ID | grep -c FINISHED`" == "0" ] ; then
+    echo "image composition failed"
+    echo "check 'composer-cli compose status'"
+    exit 1
+fi
+
+
+# saving image to web folder
+echo "saving the new image to web folder "
```
Comment (masayag) [inline r4e-img-upgrade-iso.sh:55]: pls include the name of that folder in this message, so the user can refer to it from the kickstart file / beaker job if/when needed. Ideally, we could provide here a link to the http server that exposes the new commit, so the user can just copy it and paste it on the device ostree conf.

**Review by masayag — CHANGES_REQUESTED** [review]

File: r4e-img-upgrade-iso.sh (hunk 7 above)
Comment (bardielle) [reply r4e-img-upgrade-iso.sh:1]: in the upgrade flow we don't need an iso
and if this is a new edge creation the first part would build a most updated image (checked it)

File: r4e-img-upgrade-iso.sh (hunk 4 above)
Comment (bardielle) [reply r4e-img-upgrade-iso.sh:39]: as it wrote here... 
https://github.com/ydayagi/r4e/blob/bd50530b5956e506e7ae168dd8770ff6d6119381/r4e-img-upgrade-iso.sh#L8

File: greenboot-health-check.sh (hunk 1 above)
Comment (bardielle) [reply greenboot-health-check.sh:3]: I added a relevant task to enrich all 4 parts of greenboot checks 
https://issues.redhat.com/browse/ECOPROJECT-340

File: r4e-img-upgrade-iso.sh (hunk 3 above)
Comment (bardielle) [reply r4e-img-upgrade-iso.sh:30]: I can add it to the README file 
First I did by using separated files and templates but it's simpler to use the same file

File: greenboot-health-check.sh [hunk 9]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+if [ -x /usr/libexec/yggdrasil/device-worker ]; then
```
Comment (masayag) [inline greenboot-health-check.sh:3]: WDYT of adding a check for the yggdrasil service ?

File: r4e-img-upgrade-iso.sh [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+BLUEPRINT_FILE=blueprint.toml
+BLUEPRINT_NAME=${IMAGE_NAME}
+
+
+# create a new blueprint
+composer-cli blueprints push $BLUEPRINT_FILE
+
+# validate blueprint
+echo "validating blueprint $BLUEPRINT_NAME"
+composer-cli blueprints depsolve $BLUEPRINT_NAME
```
Comment (masayag) [inline r4e-img-upgrade-iso.sh:39]: if dependencies are missing will it this command fail and terminate the entire script ?

File: r4e-img-upgrade-iso.sh [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+# build image using the composer based on the parent commit create a new commit- first find the parent commit hash in http://<host-ip>/<image-name>/compose.json under "ostree-commit"
+echo "creating a new commit from parrent $OSTREE_COMMIT using blueprint $BLUEPRINT_NAME"
+composer-cli compose start-ostree $BLUEPRINT_NAME rhel-edge-commit --parent $OSTREE_COMMIT > temp_upgrade.out
```
Comment (masayag) [inline r4e-img-upgrade-iso.sh:43]: using `composer-cli -j` and `jq` we can rely on the output format to analyze the build-id which is more expected than parsing output of a command.
PTAL https://github.com/ydayagi/r4e/pull/7/files#diff-76f2becada98489f751e9e666d02e9c81a6e92ea589750a299c8d019bfd54a43R98

File: greenboot-health-check.sh (hunk 9 above)
Comment (bardielle) [reply greenboot-health-check.sh:3]: Sure - I'm working on it : https://issues.redhat.com/browse/ECOPROJECT-340

File: r4e-img-upgrade-iso.sh (hunk 10 above)
Comment (bardielle) [reply r4e-img-upgrade-iso.sh:39]: No thanks I will change it

**Review by masayag — APPROVED** [review]

**Pull request merged**